cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc

```diff
@@ -1,17 +1,184 @@
 #include "tiledb_cxx_array_util.h"
-//#include "tiledb_cxx_json.hpp"
-//#include "tiledb_cxx_string_util.h"
+
 
 namespace tiledb {
 
-	std::string ArrayUtil::get_tiledb_version()
-	{
-		std::vector<int> t = tiledb::version();
-		return std::to_string(t[0]) + "." + std::to_string(t[1]) + "." + std::to_string(t[2]);
-	}//std::string ArrayUtil::get_tiledb_version()
+std::string ArrayUtil::get_tiledb_version()
+{
+  std::vector<int> t = tiledb::version();
+  return std::to_string(t[0]) + "." + std::to_string(t[1]) + "." + std::to_string(t[2]);
+}//std::string ArrayUtil::get_tiledb_version()
+
+int ArrayUtil::export_file_to_path(const std::string& file_uri, const std::string& output_path, uint64_t buffer_size, const std::shared_ptr<tiledb::Context>& ctx)
+{
+  const std::string METADATA_SIZE_KEY = "file_size";
+  const std::string FILE_ATTRIBUTE_NAME = "contents";
+  const std::string METADATA_ORIGINAL_FILE_NAME = "original_file_name";
+  try {
+
+    std::shared_ptr<Array> array = std::shared_ptr<Array>(new Array(ctx, file_uri, tiledb::QueryType::TILEDB_READ));
+
+    //get file_size from metadata
+    uint32_t v_num;
+    const void* v_r;
+    tiledb::DataType file_size_datatype = tiledb::DataType::TILEDB_UINT64;
+    array->get_metadata(METADATA_SIZE_KEY, &file_size_datatype, &v_num, &v_r);
+    uint64_t file_size = *((const int64_t*)v_r);
+    if(buffer_size==0) {
+      buffer_size = file_size;
+    }
+    if (buffer_size <= 0) {
+      std::cout << "ArrayUtil::export_file_to_path, wrong buffer_size!" << std::endl;
+      return -1;
+    }
+
+    std::string output_file = output_path;
+    if(output_file.empty()) {
+      tiledb::DataType file_name_datatype = tiledb::DataType::TILEDB_STRING_ASCII;
+      uint32_t name_size = 0;
+      const char* original_name;
+      array->get_metadata(METADATA_ORIGINAL_FILE_NAME,&file_name_datatype,&name_size,reinterpret_cast<const void**>(&original_name));
+      output_file = std::string(original_name, name_size);
+    }
+    
+    std::shared_ptr<tiledb::VFS> vfs = std::shared_ptr<tiledb::VFS>(new tiledb::VFS(ctx));
+    tiledb::VFS::filebuf filebuf = tiledb::VFS::filebuf(vfs);
+    if (filebuf.open(output_file, std::ios::out) == nullptr) {
+      std::cout << "ArrayUtil::export_file_to_path, can not open output_path:" << output_path << " for writing!" << std::endl;
+      return -1;
+    }
+    std::ostream os(&filebuf);   
+    
+    uint64_t offset = 0;
+    std::array<uint64_t, 2> subarray = { offset, offset + buffer_size - 1 };
```

Comment:
Setting this subarray means we will only read at-most `buffer_size`, rather than the whole file -- we need to query for `file_size`. We handle reading in chunks in the `while` loop below.

Reply:
I'm not sure what the goal is with `offset` here, seems unnecessary/cruft?

Reply:
changed the subarray to {0,file_size-1} and removed offset

Reply:
Thanks